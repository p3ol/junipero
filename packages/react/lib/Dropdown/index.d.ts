import { Placement } from '@floating-ui/react-dom-interactions';
import {
  Props as clickProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useClick';
import {
  Props as dismissProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useDismiss';
import {
  Props as hoverProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useHover';
import React, { MutableRefObject } from 'react';

export declare type DropdownRef = {
  opened: Boolean;
  toggle: () => void;
  open: () => void;
  close: () => void;
  isJunipero: Boolean;
  innerRef: MutableRefObject<any>;
};

declare interface DropdownProps extends React.ComponentPropsWithRef<any> {
  className?: String;
  container?: String | React.ReactNode;
  floatingOptions?: Object;
  clickOptions?: clickProps;
  hoverOptions?: hoverProps;
  dismissOptions?: dismissProps;
  openend?: Boolean;
  placement?: Placement;
  trigger?: 'click' | 'hover' | 'manual';
  onToggle?: (props: { opened: Boolean }) => void;
  disabled?: Boolean;
  ref?: MutableRefObject<DropdownRef | undefined>;
}

declare function Dropdown(props: DropdownProps): JSX.Element;

export default Dropdown;
