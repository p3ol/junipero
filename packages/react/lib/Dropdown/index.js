import {
  forwardRef,
  useCallback,
  useImperativeHandle,
  useRef,
  useReducer,
  useEffect,
} from 'react';
import { classNames, mockState } from '@junipero/core';
import {
  offset,
  autoUpdate,
  flip,
  shift,
  safePolygon,
  useFloating,
  useInteractions,
  useDismiss,
  useClick,
  useHover,
} from '@floating-ui/react-dom-interactions';
import PropTypes from 'prop-types';

import { DropdownContext } from '../contexts';

const Dropdown = forwardRef(({
  className,
  container,
  disabled,
  floatingOptions,
  clickOptions,
  hoverOptions,
  dismissOptions,
  opened = false,
  placement = 'bottom-start',
  trigger = 'click',
  onToggle,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    opened,
  });
  const { x, y, reference, floating, strategy, context } = useFloating({
    open: state.opened,
    onOpenChange: (...args) => onOpenChange(...args),
    placement,
    strategy: 'fixed',
    whileElementsMounted: autoUpdate,
    ...floatingOptions || {},
    middleware: floatingOptions?.middleware || [
      offset(10),
      flip(),
      shift(),
    ],
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context, {
      enabled: trigger === 'click',
      ...clickOptions || {},
    }),
    useHover(context, {
      enabled: trigger === 'hover',
      handleClose: safePolygon(),
      ...hoverOptions || {},
    }),
    useDismiss(context, {
      enabled: trigger === 'click',
      ...dismissOptions || {},
    }),
  ]);

  useEffect(() => {
    if (disabled && state.opened) {
      close();
    }
  }, [disabled]);

  useImperativeHandle(ref, () => ({
    opened: state.opened,
    toggle,
    open,
    close,
    isJunipero: true,
    innerRef,
  }));

  const onOpenChange = o => {
    if (disabled && o) {
      return;
    }

    dispatch({ opened: o });
    onToggle?.({ opened: o });
  };

  const toggle = () => {
    if (disabled) {
      return;
    }

    if (state.opened) {
      close();
    } else {
      open();
    }
  };

  const open = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true });
    onToggle?.({ opened: true });
  };

  const close = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  };

  const getContext = useCallback(() => ({
    opened: state.opened,
    container,
    x,
    y,
    reference,
    floating,
    strategy,
    toggle,
    open,
    close,
    getReferenceProps,
    getFloatingProps,
  }), [
    state.opened,
    x,
    y,
    reference,
    floating,
    strategy,
  ]);

  return (
    <DropdownContext.Provider value={getContext()}>
      <div
        { ...rest }
        className={classNames(
          'junipero dropdown',
          {
            opened: state.opened,
            disabled,
          },
          className
        )}
        ref={innerRef}
      />
    </DropdownContext.Provider>
  );
});

Dropdown.displayName = 'Dropdown';
Dropdown.propTypes = {
  clickOptions: PropTypes.object,
  container: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.object,
    PropTypes.node,
  ]),
  disabled: PropTypes.bool,
  dismissOptions: PropTypes.object,
  floatingOptions: PropTypes.object,
  hoverOptions: PropTypes.object,
  opened: PropTypes.bool,
  placement: PropTypes.string,
  trigger: PropTypes.oneOf(['click', 'hover', 'manual']),
  onToggle: PropTypes.func,
};

export default Dropdown;
