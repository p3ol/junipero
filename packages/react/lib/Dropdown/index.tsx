import {
  type RefObject,
  type ReactElement,
  useCallback,
  useImperativeHandle,
  useRef,
  useReducer,
  useEffect,
} from 'react';
import { classNames, omit, mockState } from '@junipero/core';
import {
  type UseClickProps,
  type UseDismissProps,
  type UseHoverProps,
  type Placement,
  type OpenChangeReason,
  type UseFloatingOptions,
  type Boundary,
  offset,
  autoUpdate,
  flip,
  shift,
  safePolygon,
  useFloating,
  useInteractions,
  useDismiss,
  useClick,
  useHover,
} from '@floating-ui/react';

import type { JuniperoRef, SpecialComponentPropsWithRef } from '../types';
import { DropdownContext, type DropdownContextType } from '../contexts';
import AccessibilityStore from '../AccessibilityStore';

export declare interface DropdownRef extends JuniperoRef {
  opened: boolean;
  toggle(): void;
  open(): void;
  close(): void;
  innerRef: RefObject<HTMLDivElement>;
}

export declare interface DropdownProps extends Omit<
  SpecialComponentPropsWithRef<'div', DropdownRef>,
  'onToggle'
>{
  clickOptions?: UseClickProps;
  container?: string | ReactElement | DocumentFragment | HTMLElement;
  disabled?: boolean;
  dismissOptions?: UseDismissProps;
  floatingOptions?: UseFloatingOptions & {
    boundary?: Boundary;
  };
  hoverOptions?: UseHoverProps;
  opened?: boolean;
  placement?: Placement;
  trigger?: 'click' | 'hover' | 'manual';
  withAccessibility?: boolean;
  onToggle?(props: { opened: boolean }): void;
}

export declare interface DropdownState {
  opened: boolean;
  visible: boolean;
  highlightedOptionId: string | null;
}

const Dropdown = ({
  ref,
  className,
  container,
  disabled,
  floatingOptions,
  clickOptions,
  hoverOptions,
  dismissOptions,
  opened = false,
  placement = 'bottom-start',
  trigger = 'click',
  withAccessibility = true,
  onToggle,
  ...rest
}: DropdownProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const [state, dispatch] = useReducer(mockState<DropdownState>, {
    opened: opened ?? false,
    visible: opened ?? false,
    highlightedOptionId: null,
  });
  const { x, y, refs, strategy, context } = useFloating({
    open: state.opened,
    onOpenChange: (...args) => onOpenChange(...args),
    placement,
    strategy: 'fixed',
    whileElementsMounted: autoUpdate,
    ...omit(floatingOptions || {}, ['boundary', 'middleware']),
    middleware: floatingOptions?.middleware || [
      offset(10),
      flip({
        boundary: floatingOptions?.boundary ||
          floatingOptions?.elements?.reference,
      }),
      shift({
        boundary: floatingOptions?.boundary ||
          floatingOptions?.elements?.reference,
      }),
    ],
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context, {
      enabled: trigger === 'click',
      ...clickOptions || {},
    }),
    useHover(context, {
      enabled: trigger === 'hover',
      handleClose: safePolygon(),
      ...hoverOptions || {},
    }),
    useDismiss(context, {
      enabled: trigger === 'click',
      ...dismissOptions || {},
    }),
  ]);

  useEffect(() => {
    if (disabled && state.opened) {
      close();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [disabled]);

  useImperativeHandle(ref, () => ({
    opened: state.opened,
    toggle,
    open,
    close,
    isJunipero: true,
    innerRef,
  }));

  const onOpenChange = (o: boolean, _: Event, __: OpenChangeReason) => {
    if (disabled && o) {
      return;
    }

    dispatch({ opened: o, visible: true });
    onToggle?.({ opened: o });
  };

  const open = useCallback(() => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true, visible: true });
    onToggle?.({ opened: true });
  }, [disabled, onToggle]);

  const close = useCallback(() => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  }, [disabled, onToggle]);

  const toggle = useCallback(() => {
    if (disabled) {
      return;
    }

    if (state.opened) {
      close();
    } else {
      open();
    }
  }, [disabled, state.opened, open, close]);

  const onAnimationExit = useCallback(() => {
    dispatch({ visible: false });
  }, []);

  const getContext = useCallback<() => DropdownContextType>(() => ({
    opened: state.opened,
    visible: state.visible,
    container,
    x,
    y,
    refs,
    strategy,
    toggle,
    open,
    close,
    getReferenceProps,
    getFloatingProps,
    onAnimationExit,
  }), [
    state.opened, state.visible,
    x, y, refs, strategy,
    toggle, open, close,
    getReferenceProps, getFloatingProps, onAnimationExit, container,
  ]);

  const dropdown = (
    <DropdownContext.Provider value={getContext()}>
      <div
        { ...rest }
        className={classNames(
          'junipero dropdown',
          {
            opened: state.opened,
            disabled,
          },
          className
        )}
        ref={innerRef}
      />
    </DropdownContext.Provider>
  );

  return withAccessibility ? (
    <AccessibilityStore>{ dropdown }</AccessibilityStore>
  ) : dropdown;
};

Dropdown.displayName = 'Dropdown';

export default Dropdown;
