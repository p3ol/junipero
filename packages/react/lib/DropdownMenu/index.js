import { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, ensureNode } from '@junipero/core';
import { createPortal } from 'react-dom';

import { useDropdown } from '../hooks';

const DropdownMenu = forwardRef(({
  children,
  className,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const {
    x,
    y,
    floating,
    strategy,
    opened,
    getFloatingProps,
    container,
  } = useDropdown();

  useImperativeHandle(ref, () => ({
    innerRef,
    isJunipero: true,
  }));

  if (!opened) {
    return null;
  }

  const content = (
    <div
      { ...rest }
      ref={floating}
      style={{
        position: strategy,
        top: y ?? 0,
        left: x ?? 0,
        ...rest.style || {},
      }}
      className={classNames('junipero dropdown-menu', className)}
      { ...getFloatingProps() }
    >
      <ul className="menu-inner">
        { children }
      </ul>
    </div>
  );

  return container ? createPortal(content, ensureNode(container)) : content;
});

DropdownMenu.displayName = 'DropdownMenu';

export default DropdownMenu;
