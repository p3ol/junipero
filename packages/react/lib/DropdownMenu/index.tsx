import {
  type RefObject,
  type ReactNode,
  useRef,
  useImperativeHandle,
  useEffect,
} from 'react';
import { classNames, ensureNode } from '@junipero/core';
import { createPortal } from 'react-dom';

import type { TransitionProps } from '../Transition';
import type { JuniperoRef, SpecialComponentPropsWithRef } from '../types';
import { useAccessibility, useDropdown } from '../hooks';

export declare interface DropdownMenuRef extends JuniperoRef {
  innerRef: RefObject<HTMLDivElement>;
}

export declare interface DropdownMenuProps
  extends SpecialComponentPropsWithRef<'div', DropdownMenuRef> {
  apparition?: string;
  autoFocus?: boolean;
  animate?(
    menu: ReactNode,
    opts: {
      opened: boolean;
    } & Partial<TransitionProps>
  ): ReactNode;
}

export const DropdownMenu = ({
  ref,
  apparition,
  children,
  className,
  autoFocus = true,
  animate,
  ...rest
}: DropdownMenuProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const {
    x,
    y,
    refs,
    strategy,
    opened,
    visible,
    container,
    getFloatingProps,
    onAnimationExit,
  } = useDropdown();
  const { onKeyDown, currentlyFocusedElement, toggleId } = useAccessibility();

  useEffect(() => {
    if (innerRef.current && autoFocus) {
      innerRef.current.focus();
    }
  }, [opened, autoFocus]);

  useImperativeHandle(ref, () => ({
    innerRef,
    isJunipero: true,
  }));

  if (!opened && !animate) {
    return null;
  }

  const menu = (
    <ul className="menu-inner">
      { children }
    </ul>
  );

  const content = (
    <div
      { ...rest }
      ref={r => {
        refs.setFloating?.(r);
        innerRef.current = r;
      }}
      style={{
        position: strategy,
        top: y ?? 0,
        left: x ?? 0,
        ...rest.style || {},
      }}
      className={classNames('junipero dropdown-menu', className)}
      { ...getFloatingProps() }
      tabIndex={0}
      onKeyDown={onKeyDown}
      role="listbox"
      aria-labelledby={toggleId}
      aria-activedescendant={currentlyFocusedElement}
    >
      { animate ? animate(menu, {
        opened,
        onExited: onAnimationExit,
      }) : menu }
    </div>
  );

  return opened || (animate && visible) || apparition === 'css'
    ? container ? createPortal(content, ensureNode(container))
      : content
    : null;
};

DropdownMenu.displayName = 'DropdownMenu';

export default DropdownMenu;
