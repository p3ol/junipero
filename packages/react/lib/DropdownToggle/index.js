import {
  Children,
  forwardRef,
  cloneElement,
  useImperativeHandle,
  useRef,
} from 'react';
import { classNames } from '@junipero/core';

import { useDropdown } from '../hooks';

const DropdownToggle = forwardRef(({
  children,
}, ref) => {
  const innerRef = useRef();
  const { opened, refs, getReferenceProps } = useDropdown();

  useImperativeHandle(ref, () => ({
    innerRef,
    isJunipero: true,
  }));

  const child = Children.only(children);

  return cloneElement(child, {
    className: classNames(child.props.className, 'dropdown-toggle', { opened }),
    ref: r => {
      innerRef.current = r?.isJunipero ? r.innerRef.current : r;
      refs.setReference(r?.isJunipero ? r.innerRef.current : r);
    },
    ...getReferenceProps({ onClick: child.props.onClick }),
  });
});

DropdownToggle.displayName = 'DropdownToggle';
DropdownToggle.propTypes = {};

export default DropdownToggle;
