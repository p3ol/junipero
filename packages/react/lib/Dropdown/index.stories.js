import { action } from '@storybook/addon-actions';
import { CSSTransition } from 'react-transition-group';

import Button from '../Button';
import Dropdown from './index';
import DropdownToggle from '../DropdownToggle';
import DropdownMenu from '../DropdownMenu';
import DropdownItem from '../DropdownItem';

export default { title: 'junipero/Dropdown' };

export const basic = () => (
  <Dropdown onToggle={action('toggle')}>
    <DropdownToggle>Click me</DropdownToggle>
    <DropdownMenu>
      <DropdownItem><a>Menu item 1</a></DropdownItem>
      <DropdownItem><a>Menu item 2</a></DropdownItem>
      <DropdownItem><a>Menu item 3</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const opened = () => (
  <Dropdown opened={true} onToggle={action('toggle')}>
    <DropdownToggle>Click me</DropdownToggle>
    <DropdownMenu>
      <DropdownItem><a>Menu item 1</a></DropdownItem>
      <DropdownItem><a>Menu item 2</a></DropdownItem>
      <DropdownItem><a>Menu item 3</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const withButton = () => (
  <Dropdown onToggle={action('toggle')}>
    <DropdownToggle>
      <Button className="primary">Click me</Button>
    </DropdownToggle>
    <DropdownMenu>
      <DropdownItem><a>Menu item 1</a></DropdownItem>
      <DropdownItem><a>Menu item 2</a></DropdownItem>
      <DropdownItem><a>Menu item 3</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const withCssApparition = () => (
  <Dropdown onToggle={action('toggle')}>
    <DropdownToggle>Click me</DropdownToggle>
    <DropdownMenu apparition="css">
      <DropdownItem><a>Menu item 1</a></DropdownItem>
      <DropdownItem><a>Menu item 2</a></DropdownItem>
      <DropdownItem><a>Menu item 3</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const animated = () => (
  <Dropdown
    onToggle={action('toggle')}
  >
    <DropdownToggle>Click to animate me</DropdownToggle>
    <DropdownMenu
      animate={(menu, { opened }) => (
        <CSSTransition
          in={opened}
          mountOnEnter={true}
          unmountOnExit={true}
          timeout={300}
          classNames="slide-in-up-dropdown"
          children={menu}
        />
      )}
    >
      <DropdownItem><a>Menu item 1</a></DropdownItem>
      <DropdownItem><a>Menu item 2</a></DropdownItem>
      <DropdownItem><a>Menu item 3</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);
