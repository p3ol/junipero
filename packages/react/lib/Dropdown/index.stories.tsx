import { slideInDownMenu } from '@junipero/transitions';
import { action } from 'storybook/actions';

import DropdownToggle from '../DropdownToggle';
import DropdownMenu from '../DropdownMenu';
import DropdownGroup from '../DropdownGroup';
import Tag from '../Tag';
import Dropdown from '.';
import DropdownItem from '../DropdownItem';

export default { title: 'react/Dropdown' };

export const Basic = () => (
  <Dropdown>
    <DropdownToggle><Tag tabIndex={0}>Open me</Tag></DropdownToggle>
    <DropdownMenu>
      <DropdownItem><a onClick={action('item1')}>Item 1</a></DropdownItem>
      <DropdownItem><a onClick={action('item2')}>Item 2</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const WithAccessibleIds = () => (
  <Dropdown id="my-dropdown">
    <DropdownToggle><Tag>Open me</Tag></DropdownToggle>
    <DropdownMenu id="my-dropdown-menu">
      <DropdownItem><a onClick={action('item1')}>Item 1</a></DropdownItem>
      <DropdownItem><a onClick={action('item2')}>Item 2</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const Hover = () => (
  <Dropdown trigger="hover">
    <DropdownToggle><Tag>Open me</Tag></DropdownToggle>
    <DropdownMenu>
      <DropdownItem><a>Item 1</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const WithGroups = () => (
  <Dropdown>
    <DropdownToggle><Tag>Open me</Tag></DropdownToggle>
    <DropdownMenu>
      <DropdownGroup title="Test group">
        <DropdownItem><a>Item 1</a></DropdownItem>
      </DropdownGroup>
      <DropdownGroup title="Another group">
        <DropdownItem><a>Item 2</a></DropdownItem>
      </DropdownGroup>
    </DropdownMenu>
  </Dropdown>
);

export const Animated = () => (
  <Dropdown>
    <DropdownToggle><Tag>Open me</Tag></DropdownToggle>
    <DropdownMenu animate={slideInDownMenu}>
      <DropdownItem><a>Item 1</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);

export const AnimatedButStillInDom = () => (
  <Dropdown>
    <DropdownToggle><Tag>Open me</Tag></DropdownToggle>
    <DropdownMenu animate={slideInDownMenu} apparition="css">
      <DropdownItem><a>Item 1</a></DropdownItem>
    </DropdownMenu>
  </Dropdown>
);
