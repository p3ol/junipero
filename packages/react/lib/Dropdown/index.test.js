import { fireEvent, render } from '@testing-library/react';

import DropdownToggle from '../DropdownToggle';
import DropdownMenu from '../DropdownMenu';
import Dropdown from './index';

describe('<Dropdown />', () => {
  it('should render', () => {
    const { getByText, container, unmount } = render(
      <Dropdown>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu>
          <li>Item 1</li>
        </DropdownMenu>
      </Dropdown>
    );
    expect(container).toMatchSnapshot();

    fireEvent.click(getByText('Click me'));

    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should auto close when disabled prop changes', () => {
    const { container, unmount, rerender, getByText } = render(
      <Dropdown opened={true}>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu>
          <li>Item 1</li>
        </DropdownMenu>
      </Dropdown>
    );

    expect(container).toMatchSnapshot();

    rerender(
      <Dropdown disabled>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu>
          <li>Item 1</li>
        </DropdownMenu>
      </Dropdown>
    );

    fireEvent.click(getByText('Click me'));

    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should allow to render menu inside another element', () => {
    const { container, unmount } = render(
      <>
        <div id="menu-container" />
        <Dropdown container="#menu-container" opened={true}>
          <DropdownToggle><span>Click me</span></DropdownToggle>
          <DropdownMenu>
            <li>Item 1</li>
          </DropdownMenu>
        </Dropdown>
      </>
    );

    expect(container).toMatchSnapshot();
    unmount();
  });
});
