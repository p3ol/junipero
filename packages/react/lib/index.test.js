import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
