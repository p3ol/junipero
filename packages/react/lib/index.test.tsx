import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
