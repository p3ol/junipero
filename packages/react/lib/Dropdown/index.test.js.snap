// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should allow to render menu inside another element 1`] = `
<div>
  <div
    id="menu-container"
  >
    <div
      class="junipero dropdown-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <li>
          Item 1
        </li>
      </ul>
    </div>
  </div>
  <div
    class="junipero dropdown opened"
  >
    <span
      class="dropdown-toggle opened"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`<Dropdown /> should auto close when disabled prop changes 1`] = `
<div>
  <div
    class="junipero dropdown opened"
  >
    <span
      class="dropdown-toggle opened"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <li>
          Item 1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> should auto close when disabled prop changes 2`] = `
<div>
  <div
    class="junipero dropdown disabled"
  >
    <span
      class="dropdown-toggle"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`<Dropdown /> should render 1`] = `
<div>
  <div
    class="junipero dropdown"
  >
    <span
      class="dropdown-toggle"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`<Dropdown /> should render 2`] = `
<div>
  <div
    class="junipero dropdown opened"
  >
    <span
      class="dropdown-toggle opened"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <li>
          Item 1
        </li>
      </ul>
    </div>
  </div>
</div>
`;
